typing/typeopt.ml

```diff
@@ -27,11 +27,9 @@ let scrape_ty env ty =
       let ty = Ctype.expand_head_opt env (Ctype.correct_levels ty) in
       begin match get_desc ty with
       | Tconstr (p, _, _) ->
-          begin match Env.find_type p env with
-          | {type_kind = ( Type_variant (_, Variant_unboxed)
-          | Type_record (_, Record_unboxed _) ); _} ->
-            Ctype.get_unboxed_type_representation env ty
-          | _ -> ty
+          begin match decl_is_unboxed (Env.find_type p env) with
+          | Some _ -> Ctype.get_unboxed_type_representation env ty
```

Comment:
It's slightly frustrating that we cannot easily use the argument of the `Some` constructor to continue the unboxed-type-representation computation, as we know that `Ctype.get_unboxed_type_representation` is going to recompute this argument immediately. But there isn't an easy way (we don't want to duplicate the logic of that case inside the get_unboxed_... function that handles Tpoly etc.), so I think that the current code here is fine.

Reply:
I agree this is unfortunate (though perhaps this PR mainly makes it more evident, not much worse than it was).

One could turn `get_unboxed_type_representation` into a pair of mutually recursive functions, where just the innermost part (the three lines beginning `let ty2 = ...`) is its own function, and then add the new function to the mli for `typeopt` to call.  I think that the resulting code would be confusing (it's not even clear to me what a sensible name for the new function would be), so my instinct is not to do this, but I'm happy to try if you think the performance improvement is worth it.

Reply:
I agree that leaving things unboxed for now is fine. (I was going to say that I'm not worried the least about performance, it was about having more natural code logic by avoiding seemingly redundant computation. In fact *some* of these functions might be performance-critical, `get_unboxed_type_repr` and `Typeopt.scrape` may matter -- the other certainly don't. But I think that they matter in other ways than slightly redundant computations, on the value head in the non-common case.)